### PR TITLE
WIP [JENKINS-40665] Pipeline support for Maven SettingsProvider & GlobalSettingsProvider

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
Support Maven Settings Provider and Maven Global Settings Provider for pipeline and make it work with `withMaven(){}`

See
* [JENKINS-40665 Maven SettingsProvider and GlobalSettingsProvider should support Pipeline jobs](https://issues.jenkins-ci.org/browse/JENKINS-40665)
* [JENKINS-39407 Use system default Maven, Maven Settings if not specified](https://issues.jenkins-ci.org/browse/JENKINS-39407)
* PR https://github.com/jenkinsci/jenkins/pull/2678
* PR https://github.com/jenkinsci/pipeline-maven-plugin/pull/31
* PR https://github.com/jenkinsci/config-file-provider-plugin/pull/25

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
@jglick could you have a look at this PR?

Reference: [JENKINS-39407 Use system default Maven, Maven Settings if not specified](https://issues.jenkins-ci.org/browse/JENKINS-39407)

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: I think the newly introduced method should support propagation of non-Runtime exceptions if there is a real need to propagate them

File: core/src/main/java/jenkins/mvn/SettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/SettingsProvider.java]: 🐛 The implementation should check `instanceof AbstractBuild` and invoke old method by default. Otherwise it should return null and maybe log warning

File: core/src/main/java/jenkins/mvn/SettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/SettingsProvider.java]: I would expect this method to be deprecated

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
IMHO having two unrelated implementation is a very confusing way of extending API. IMHO the old method needs to be properly deprecated as we do in other such cases.

Another 🐛 is for throwing undocumented Runtime exceptions in new APIs

**Comment by cyrille-leclerc** [comment]
@oleg-nenashev PR updated

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: Add `@since TODO`

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: It may break the summary line in Javadoc. I would keep the first entry without paragraph tags

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: I would rather call to the old `supplySettings` method if that method is overridden && the run is AbstractBuild. Otherwise I would print warning and return null. In such case new API won't blow up if somebody invokes it for non-updated providers

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: Without the fix proposed above it would be a wrong guidance since the provider implementations may not support the legacy API

File: core/src/main/java/jenkins/mvn/SettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/SettingsProvider.java]: same comment about pluggability as above

File: core/src/main/java/jenkins/mvn/SettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/SettingsProvider.java]: same comment as above

File: core/src/main/java/jenkins/mvn/SettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/SettingsProvider.java]: Should also have the `@Deprecated` annotation

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
It's getting closer, but I still feel that throwing errors from unimplemented new methods is not the best behavior we could have. The tests are also missing.

File: core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java
Comment (KostyaSha) [inline core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java]: Breaks binary compatibility.

**Review by KostyaSha — CHANGES_REQUESTED** [review]
binary compatibility breakage

File: core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java]: @KostyaSha This is a method with `@Override`. So it does not break the binary compatibility

Comment (cyrille-leclerc) [inline core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java]: Thanks for reviewing @KostyaSha. Do you agree with @oleg-nenashev's statement? Do you agree that this PR does not break binary compatibility as this `FilePath supplySettings(AbstractBuild, TaskListener)` method is still declared in the parent class  GlobalSettingsProvider: 
https://github.com/cyrille-leclerc/jenkins/blob/JENKINS-40665/core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java#L57

**Comment by cyrille-leclerc** [comment]
@oleg-nenashev  thanks for reviewing, I just had a little baby girl, I will be a bit slower to update the PR according to your comments.
@KostyaSha I'll wait for your reply on the binary compatibility issue

**Comment by KostyaSha** [comment]
Lost override, dismissing review

File: core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java
Comment (KostyaSha) [inline core/src/main/java/jenkins/mvn/DefaultGlobalSettingsProvider.java]: Yes, i missed override side.

**Comment by cyrille-leclerc** [comment]
@oleg-nenashev PR updated. The tests would be the last bit.

**Comment by cyrille-leclerc** [comment]
The tests in error don't seem to be related to the content of the PR:

```
Tests in error: 
jenkins.security.Security232Test.commonsCollections1(jenkins.security.Security232Test)
  Run 1: Security232Test.commonsCollections1:85 » Socket Connection reset
  Run 2: Security232Test.commonsCollections1:85 » Socket Connection reset
  Run 3: Security232Test.commonsCollections1:85 » Socket Connection reset
  Run 4: Security232Test.commonsCollections1:85 » Socket Connection reset
  Run 5: Security232Test.commonsCollections1:85 » Socket Connection reset
```

**Comment by daniel-beck** [comment]
@cyrille-leclerc 

> don't seem to be related

Known flaky test.

**Comment by KostyaSha** [comment]
> we have two for redundancy,

And no quorum

**Comment by oleg-nenashev** [comment]
It was my AI to review it. Will do it ASAP

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me 🐝 
Ideally it requires some testing to confirm non-AbstractBuilds do not blow up due to AbstractMethodError with this change.

**Comment by cyrille-leclerc** [comment]
@oleg-nenashev thanks, I have tested with:

* config-file-provider-plugin PR https://github.com/jenkinsci/config-file-provider-plugin/pull/25 /  [JENKINS-40665 Maven SettingsProvider and GlobalSettingsProvider should support Pipeline jobs](https://issues.jenkins-ci.org/browse/JENKINS-40665)
* pipeline-maven-plugin PR https://github.com/jenkinsci/pipeline-maven-plugin/pull/15 / [JENKINS-39407 Use system default Maven, Maven Settings if not specified](https://issues.jenkins-ci.org/browse/JENKINS-39407)

**Comment by oleg-nenashev** [comment]
@KostyaSha @jglick WDYT?

File: core/src/main/java/jenkins/mvn/FilePathGlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/FilePathGlobalSettingsProvider.java]: Really you want to treat this as a path on master and not the agent? Seems like an unwanted “feature”.

File: core/src/main/java/jenkins/mvn/FilePathGlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/FilePathGlobalSettingsProvider.java]: Note that for Pipeline this will be the build’s global environment—it will not take into consideration `node`, `withEnv`, etc. If you actually want to use a local environment, you must pass in an `EnvVars`.

File: core/src/main/java/jenkins/mvn/FilePathSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/FilePathSettingsProvider.java]: Consolidate code duplicated with `FilePathGlobalSettingsProvider` into a helper function.

File: core/src/main/java/jenkins/mvn/FilePathGlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/FilePathGlobalSettingsProvider.java]: Delete—I thought the whole reason to allow the method to throw `IOException | InterruptedException` was so that we would need to wrap in a `RuntimeException`.

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: Uh, no it does not.

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: use `@link` please

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: use `@code`

File: core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/GlobalSettingsProvider.java]: just “build” suffices to refer casually to a `Run`

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/mvn/SettingsProvider.java
Comment (jglick) [inline core/src/main/java/jenkins/mvn/SettingsProvider.java]: Then you should do so from `Maven.perform`, right?

**Comment by oleg-nenashev** [comment]
@cyrille-leclerc Any feedback regarding https://github.com/jenkinsci/jenkins/pull/2678#discussion_r101819865 ?

@KostyaSha ping, I think your -1 is resolved

**Comment by KostyaSha** [comment]
> @KostyaSha ping, I think your -1 is resolved

> Lost override, dismissing review. Though my review isn't shown on bottom.

No blockers from my side

**Comment by oleg-nenashev** [comment]
Ready to merge then. Will integrate the changes today or tomorrow

**Comment by jglick** [comment]
No it is not ready. I discussed some problems with @cyrille-leclerc and he said he would work on it further.

**Comment by cyrille-leclerc** [comment]
The conflict is just a spelling fix in javadocs: https://github.com/jenkinsci/jenkins/commit/496703d0fe133445e10c7d8d07fa7afd351c8854#diff-a30a831cee36351bda7e7a615e9b559aL27

**Comment by cyrille-leclerc** [comment]
@jglick I have a rebased version of this branch at https://github.com/cyrille-leclerc/jenkins/commits/JENKINS-40665-rebased .
Shall I rebase on this Pull Request? Shall I close this Pull Request and create a new Pull Request with the rebased version?

**Comment by daniel-beck** [comment]
There's currently no notable "open" review feedback, so rebasing shouldn't break GitHub.  @jglick WDYT?

**Comment by cyrille-leclerc** [comment]
Here is the rebased fork, I don't see how it could cause harm to this PR but I am not very experienced in rebasing PRs: https://github.com/cyrille-leclerc/jenkins/tree/JENKINS-40665-rebased

@jglick side not: I failed to use a DockerRule in my tests, I got the following error (gist attached)

```
java.lang.IllegalArgumentException: Argument for @Nonnull parameter 'primaryView' of hudson/model/AllView.migrateLegacyPrimaryAllViewLocalizedName must not be null
```

https://gist.github.com/cyrille-leclerc/d88f4b157e85ca38afeaad4b77ae16f3

**Comment by cyrille-leclerc** [comment]
The rebase made the review of this PR hard to read. I am preparing a new PR.

**Comment by cyrille-leclerc** [comment]
Replaced by https://github.com/jenkinsci/jenkins/pull/2818

**Pull request closed without merging**
